Add trust establishment between verifier and issuer (#51)
peppelinux/openid-federation-4-wallet-architectures: openid-federation-wallet-1_0.md

~~~
**Figure 6**: Federation Entity Discovery, the Wallet Instance evaluates the trust with a Credential Verifier.

## Credential Verifiers Establishing Trust in Credential Issuers

During the Credential verification phase, Credential Verifiers SHOULD establish trust with Credential Issuers to validate the authenticity and integrity of presented digital Credentials. This trust evaluation ensures that the Credential Verifier can rely on the Credentials issued by a particular Credential Issuer and verify their validity against the Credential Issuer's cryptographic material.

The trust establishment between Credential Verifiers and Credential Issuers follows the standard OpenID Federation entity discovery mechanisms. The Credential Verifier initiates this process by fetching the Credential Issuer's Entity Configuration to obtain authority hints and construct a valid Trust Chain up to a recognized Trust Anchor.

> The [@!OpenID.Federation] specification requires that Entity Identifiers MUST be URLs using the https scheme. In various flavors of digital Credentials, the Credential Issuer can be identified in various ways. If the Credential Issuer is identified using a https URL in the digital Credential, the Credential Verifier SHOULD use that URL as the Entity Identifier of the Credential Issuer. If the Credential Issuer is identified using other methods that have 1-to-1 mapping to a https URL (e.g., `did:web`, `did:webvh`), the Credential Verifier MAY use that URL as the Entity Identifier of the Credential Issuer. Future versions of this specification may provide other mechanisms to specity the Credential Issuer's Entity Identifier in a digital credential.

Reviewer: Not sure we should rely on this 1-1 mapping to a HTTPS url. I think it would make more sense to define how to link a DID to a federation entity (using #43 and a backlink in the DID) in a generic way.

~~~
**Figure 6**: Federation Entity Discovery, the Wallet Instance evaluates the trust with a Credential Verifier.

## Credential Verifiers Establishing Trust in Credential Issuers

During the credential verification phase, Credential Verifiers SHOULD establish trust with Credential Issuers to validate the authenticity and integrity of presented digital credentials. This trust establishment process is critical to ensure that the Credential Verifier can rely on the credentials issued by a particular Credential Issuer and verify their validity against the issuer's cryptographic material.

> The trust establishment between Credential Verifiers and Credential Issuers follows the standard OpenID Federation entity discovery mechanisms. The Credential Verifier initiates this process by fetching the Credential Issuer's Entity Configuration to obtain authority hints and construct a valid Trust Chain up to a recognized Trust Anchor.

Reviewer 2: If feels a bit limiting to have federation as the only verifier->issuer trust establishment mechanism? e.g. if an mdl issuer was not part of a federation, that would mean verifiers using federation to establish a relationship with a wallet couldn't then receive mdls that originated from a non-federation supporting mdl issuer?

Reviewer 3: this draft is about using openid federation

mdl issuer using openid4vci can be part of an openid federation

if they do not use openid federation, they would use for sure X.509 PKI and the trust establishment using X.509 certificates is made using certificate chains about using a trusthworthy trust anchor

Reviewer 2: I get that - I guess what I mean is: should it be broken down into two separate sections that implementors can pick one or both to implement? e.g. I may well want to use openid federation between my wallet and verifiers even if I can't get all my issuers to support federation.

And potentially I might want to use federation between the wallets and issuers even if the verifiers don't all support federation?

> The [@!OpenID.Federation] specification requires that Entity Identifiers MUST be URLs using the https scheme. In various flavors of digital credentials, the Credential Issuer can be identified in various ways. If the Credential Issuer is identified using a https URL in the digital credential, the Credential Verifier SHOULD use that URL as the Entity Identifier of the Credential Issuer. If the Credential Issuer is identified using other methods that have 1-to-1 mapping to a https URL (e.g., `did:web`, `did:webvh`), the Credential Verifier MAY use that URL as the Entity Identifier of the Credential Issuer. Future versions of this specification may provide other mechanisms to specity the Credential Issuer's Entity Identifier in a digital credential.

Reviewer 2: I can see how this would work for (say) SD-JWT (a https url in the `iss` claim), but how would it work for mdoc? (You can use this spec for mdoc right?)

Reviewer: > please note that this is of general interest but out of the scopes of this draft

Any reason why it is out of scope of this draft? Couldn't we do something like OpenID4VCI and OpenID4VP where we define annexes for specific credential formats? I would like to have it included as part of this draft, it's one of the most important aspects of this specification.

> The [@!OpenID.Federation] specification requires that Entity Identifiers MUST be URLs using the https scheme. In various flavors of digital credentials, the Credential Issuer can be identified in various ways. If the Credential Issuer is identified using a https URL in the digital credential, the Credential Verifier SHOULD use that URL as the Entity Identifier of the Credential Issuer. If the Credential Issuer is identified using other methods that have 1-to-1 mapping to a https URL (e.g., `did:web`, `did:webvh`), the Credential Verifier MAY use that URL as the Entity Identifier of the Credential Issuer. Future versions of this specification may provide other mechanisms to specity the Credential Issuer's Entity Identifier in a digital credential.

Reviewer 2: I think we need a warning here - the fact that, e.g., `iss` in a SD-JWT VC is a https url does NOT mean that the issuer supports federation. e.g. SD-JWT VC defines a web based key resolution mechanism that uses https urls.

We should probably include a reference that URLs in the form `openid_federation:https://issuer.example.com` can be used to unambiguously indicate that federation is used?

Reviewer 3: ```suggestion
The [@!OpenID.Federation] specification requires that Entity Identifiers MUST be URLs using the https scheme. In various flavors of digital Credentials, the Credential Issuer can be identified in various ways. If the Credential Issuer is identified using a https URL in the digital Credential, the Credential Verifier MUST use that URL as the Entity Identifier of the Credential Issuer. If the Credential Issuer is identified using other methods that have 1-to-1 mapping to a URI (e.g., `did:web:`, `did:webvh:`), the Credential Verifier MAY use that URL as the Entity Identifier of the Credential Issuer.
```

@samuelmr how could an entity be identified using different identifiers, other than https URLs?
I ask since OpenID Federation only uses https URIs.

it might be useful trying to explore the feature proposed in the following issue? https://github.com/openid/federation-wallet/issues/43

Author: @peppelinux, see, for example:

- [Example 2 in VCDM 2](https://www.w3.org/TR/vc-data-model-2.0/#example-use-of-the-context-property) in which the issuer is identified by `"did:example:2g55q912ec3476eba2l9812ecbfe"`. 
- [Example 3](https://www.w3.org/TR/vc-data-model-2.0/#example-use-of-the-id-property) where the issuer is `"https://university.example/issuers/565049"`
- [Example 30](https://www.w3.org/TR/vc-data-model-2.0/#example-verifiable-credential-using-the-data-integrity-bbs-cryptosuite-with-a-base-proof) where the issuer id is `"did:web:credentials.utopia.example"`

If you got all those VCs as a verifier, you'd look for `https://university.example/issuers/565049/.well-known/openid-federation` as the Entity Identifier for the Example 3 and `https://credentials.utopia.example/.well-known/openid-federation` as the Entity Identifier for the Example 30.

You couldn't get the OpenID Federation Entity Configuration of the issuer of the Example 2 VC since that would require additional discovery mechanisms.

Author: I agree with @jogu that there's no guarantee that the Entity Configuration can be found in the Entity Identifier locations of issuers of Example 3 and Example 30.

I would like to see a more explicit reference from a credential to the issuer's Entity Identifier. I made a suggestion in #50, but got feedback (outside GitHub) that new discovery mechanisms should be avoided.

Reviewer 3: removing the requirement to use entity configuration at well-known endpoints is like removing the any other requirement of openid/oauth/connect to use their well-known endpoints

the point is about interop and discovery, see: https://github.com/openid/federation/pull/265

OpenID Federation 1.0 enables interoperability on well known approaches. Other mechanisms will be evaluated and implemented in future version of OpenID Federation (2.0 I suppose).

OpenID Federation Wallet will accomodate any advancements according to future versions of OpenID Federation, without removing its current support to OpenID Federation 1.0

> The Credential Verifier SHOULD validate the Trust Chain for the Credential Issuer using the same federation discovery process outlined in other sections. This includes:

Reviewer 2: I don't understand how the 'SHOULD' here interacts with the MUST in bullet 3? In general I'm not sure about this form of text (i.e. "outlined" and "includes") - we should either refer only to other places that fully normatively define the behaviour, or we should fully normatively define the behaviour in this section (and we should only do the latter in the case where the behaviour isn't described elsewhere already).
